libraries/AP_Common/ExpandingString.h

```diff
@@ -0,0 +1,50 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/*
+  expanding string, used for @SYS filesystem output
+ */
+
+#pragma once
+
+#include <AP_HAL/AP_HAL.h>
+
+class ExpandingString {
+public:
+
+    const char *get_string(void) const {
+        return buf;
+    }
+    uint32_t get_length(void) const {
+        return used;
+    }
+
+    // print into the string
+    void printf(const char *format, ...) FMT_PRINTF(2,3);
+
+    // append data to the string
+    void append(const char *s, uint32_t len);
```

Comment:
No reason to not make this `bool`?

Similarly with `printf`.

Doesn't need to be `WARN_IF_UNUSED`, but does give the caller the option of dealing with things.  It would allow you to set `errno` appropriately if something fails.

Reply:
I'd much rather users of this API not bother with error handling, and have the SYS filesystem code do it. The point of this API is to make things as simple as possible for adding new SYS files